### PR TITLE
feat: derive categories from state

### LINKED ISSUE #61: Refactor category domain
Categories can be derived from already existing state and will make filtering/searching easier

### PR DESCRIPTION
**Pull request by cazala**
Closes #61 

In this PR I modified the `getCategories` selector to derive the `FullCategory[]` array from the state instead of storing it in the category reducer. I also added the ability to filter by `assetPackId` and by a search string that is compared against asset name, category or tags.

I left the redux actions in case we want to wire it to the UI in the future.

### CODE REVIEW

File: src/modules/category/selectors.ts [hunk 1]
```diff
@@ -1,24 +1,74 @@
...
+import { getData as getAssetPacks } from 'modules/assetPack/selectors'
 import { getData as getAssets } from 'modules/asset/selectors'
+import { AssetState } from 'modules/asset/reducer'
+import { AssetPackState } from 'modules/assetPack/reducer'
+import { Asset } from 'modules/asset/types'
 
 export const getState: (state: RootState) => CategoryState = state => state.category
 
-export const getData: (state: RootState) => CategoryState['data'] = state => getState(state).data
+export const getSelectedAssetPackId = (state: RootState) => getState(state).selectedAssetPackId
```
Comment (nicosantangelo) [inline src/modules/category/selectors.ts]: I don't get why is this here, wouldn't this be a responsibility of `assetPack`?

Comment (nicosantangelo) [inline src/modules/category/selectors.ts]: I don't get why is this here, wouldn't this be a responsibility of assetPack?

File: src/modules/category/selectors.ts [hunk 2]
```diff
@@ -1,24 +1,74 @@
...
 
-export const isLoading: (state: RootState) => boolean = state => getState(state).loading.length > 0
+export const getSearch = (state: RootState) => getState(state).search
 
-export const getError: (state: RootState) => CategoryState['error'] = state => getState(state).error
+const isSearchResult = (asset: Asset, search: string) => {
```
Comment (nicosantangelo) [inline src/modules/category/selectors.ts]: I don't get why is this here, it's not checking _only_ the category, so wouldn't this be a responsibility of `assetPack`? or maybe `ui/search` ?

File: src/modules/category/selectors.ts [hunk 3]
```diff
@@ -1,24 +1,74 @@
...
+            }
+          }
+          // add asset to category
+          categories[asset.category].assets.push(asset)
+        }
+      }
+    }
+
+    // convert map to array
+    return Object.keys(categories).map(categoryName => categories[categoryName])
```
Comment (nicosantangelo) [inline src/modules/category/selectors.ts]: `Object.values`?

File: src/modules/category/selectors.ts (hunk 1 above)
Comment (belohlavek) [reply src/modules/category/selectors.ts]: If i recall correctly the idea was to leave the possibility to choose a single asset pack at a time, while `null` would merge all assets packs together by category.

This is one of those things that might get rolled back in the future depending on the direction we take. What do you think?

Comment (cazala) [reply src/modules/category/selectors.ts]: Maybe `category` is not the right name anymore, but the responsibility of this module is to generate the `FullCategory[]` array that we use to render the item drawer. 

We could rename it to `ui` or `sidebar` or something else, but i think the selected asset pack id and search string belong to a separate module rather that `assetPack`.

File: src/modules/category/selectors.ts (hunk 2 above)
Comment (cazala) [reply src/modules/category/selectors.ts]: https://github.com/decentraland/builder/pull/106#discussion_r253470051

File: src/modules/category/selectors.ts (hunk 1 above)
Comment (nicosantangelo) [reply src/modules/category/selectors.ts]: > We could rename it to ui or sidebar or something else, but i think the selected asset pack id and search string belong to a separate module rather that assetPack.

Agreed

> Maybe category is not the right name anymore, but the responsibility of this module is to generate the FullCategory[] array that we use to render the item drawer.

Think so, I think the only problem here (if there's one) is the name of the domain. I like both UI and sidebar

**Comment by cazala** [comment]
> I'm not sure I understand the search and selected actions, but super good otherwise

@nicosantangelo this is how the search action would work if wired to an input (not part of this PR, just to illustrate)

search

And the `selectAssetId` could be connected to a dropdown to filter by asset pack (ie. All,  Space, Western, Medieval, etc)

File: src/modules/category/selectors.ts (hunk 1 above)
Comment (cazala) [reply src/modules/category/selectors.ts]: refactored into `ui.sidebar`

**Comment by nicosantangelo** [comment]
Awesomeeeee
